src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/clusterinteraction/ClusterDetailFragment.kt

```diff
@@ -41,13 +99,122 @@ class ClusterDetailFragment : Fragment(){
+      val selectedCommand: String = commandAutoComplete.adapter.getItem(position).toString()
+      selectedCommandInfo = selectedClusterInfo.commands[selectedCommand]!!
+      selectedCommandCallback = selectedCommandInfo.commandCallbackSupplier.get()
+      populateCommandParameter(inflater, parameterList)
+      selectedCommandCallback.setCallbackDelegate(object : ClusterCommandCallback {
+        override fun onSuccess(responseValues: Map<CommandResponseInfo, Any>) {
+          showMessage("Command success")
+          // Populate UI based on response values. We know the types from CommandInfo.getCommandResponses().
+          requireActivity().runOnUiThread {
+            populateCallbackResult(
+              responseValues,
+              inflater,
+              callbackList
+            )
+          }
+          responseValues.forEach { Log.d(TAG, it.toString()) }
+        }
+
+        override fun onFailure(exception: Exception) {
+          showMessage("Command failed")
+          Log.e(TAG, exception.toString())
+        }
+      })
+    }
+  }
+
+  private fun populateCommandParameter(inflater: LayoutInflater, parameterList: LinearLayout) {
+    selectedCommandInfo.commandParameters.forEach { (paramName, paramInfo) ->
+      val param = inflater.inflate(R.layout.cluster_parameter_item, null, false) as ConstraintLayout
+      param.clusterParameterNameTv.text = "${paramName}"
+      param.clusterParameterTypeTv.text = "${paramInfo.type}"
+      parameterList.addView(param)
+    }
+  }
+
+  private fun populateCallbackResult(
+    responseValues: Map<CommandResponseInfo, Any>,
+    inflater: LayoutInflater,
+    callbackList: LinearLayout
+  ) {
+    responseValues.forEach { (variableNameType, response) ->
+      val callback =
+        inflater.inflate(R.layout.cluster_callback_item, null, false) as ConstraintLayout
+      callback.clusterCallbackNameTv.text = variableNameType.name
+      callback.clusterCallbackDataTv.text = response.toString()
+      callback.clusterCallbackTypeTv.text = variableNameType.type
+      callbackList.addView(callback)
+    }
+  }
+
+  private fun getCommandOptions(
+    clusterName: String
+  ): ArrayAdapter<String> {
+    selectedClusterInfo = clusterMap[clusterName]!!
+    val commandNameList = constructHint(selectedClusterInfo.commands)
+    return ArrayAdapter(requireContext(), android.R.layout.simple_list_item_1, commandNameList)
+  }
+
+  private fun constructHint(clusterMap: Map<String, *>): Array<String> {
+    val clusterName = mutableListOf<String>()
+    for ((name, info) in clusterMap) {
+      clusterName.add(name)
+    }
+    return clusterName.toTypedArray()
+  }
+
   override fun onStop() {
     super.onStop()
     scope.cancel()
   }
 
   companion object {
     private const val TAG = "ClusterDetailFragment"
-    fun newInstance(): ClusterDetailFragment = ClusterDetailFragment()
+    private const val ENDPOINT_ID = "endpoint_id"
+    private const val DEVICE_PTR = "device_ptr"
```

Comment:
```suggestion
    private const val ENDPOINT_ID_KEY = "endpoint_id"
    private const val DEVICE_PTR_KEY = "device_ptr"
```
Important to mark these as keys, or else it will look like ENDPOINT_ID refers to the actual endpoint id.